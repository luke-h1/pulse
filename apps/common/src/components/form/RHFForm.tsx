import camelCase from 'lodash/camelCase';
import React, {
  FormEvent,
  ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { FieldValues, useFormContext, useWatch } from 'react-hook-form';
import useMountedRef from '@common/hooks/useMountedRef';
import { VStack } from '@chakra-ui/react';
import useToggle from '@common/hooks/useToggle';
import isErrorLike from '@common/utils/isErrorLike';
import createRHFFormErrorHelper from './validation/createRHFFormErrorHelper';
import ErrorSummary from '../ErrorSummary';
import { FormIdContextProvider } from './contexts/FormIdContext';

interface Props<TFormValues extends FieldValues> {
  children: ReactNode;
  id: string;
  submitId?: string;
  showErrorSummary?: boolean;
  onSubmit: (values: TFormValues) => Promise<void>;
  showSubmitError?: boolean;
  width?: string;
}

export interface ErrorSummaryMessage {
  id: string;
  message: string;
}

export default function RHFForm<TFormValues extends FieldValues>({
  children,
  id,
  onSubmit,
  submitId = `${id}-submit`,
  showErrorSummary = true,
  showSubmitError = false,
  width,
}: Props<TFormValues>) {
  const isMounted = useMountedRef();

  const methods = useFormContext<TFormValues>();
  const { handleSubmit: RHFSubmit } = methods;
  const { isSubmitted, submitCount } = methods.formState;
  const values = useWatch<TFormValues>();
  const previousValues = useRef(values);
  const previousSubmitCount = useRef(submitCount);
  const [hasSummaryFocus, toggleSummaryFocus] = useToggle(false);

  const { getAllErrors } = createRHFFormErrorHelper({
    errors: methods.formState.errors,
    touchedFields: methods.formState.touchedFields,
    isSubmitted,
  });

  const [submitError, setSubmitError] = useState<ErrorSummaryMessage>();

  useEffect(() => {
    if (!isMounted.current) {
      // eslint-disable-next-line no-useless-return
      return;
    }
    previousValues.current = values;
  }, [isMounted, values]);

  const allErrors = useMemo(() => {
    const summaryErrors: ErrorSummaryMessage[] = Object.entries(
      getAllErrors(),
    ).map(([errorName, message]) => ({
      id: `${id}-${camelCase(errorName)}`,
      message: typeof message === 'string' ? message : '',
    }));

    return submitError && isSubmitted
      ? [...summaryErrors, submitError]
      : summaryErrors;
  }, [getAllErrors, id, submitError, isSubmitted]);

  useEffect(() => {
    if (
      (isMounted && allErrors.length,
      submitCount !== previousSubmitCount.current)
    ) {
      toggleSummaryFocus.on();
      previousSubmitCount.current = submitCount;
    }
  }, [allErrors, isMounted, submitCount, toggleSummaryFocus]);

  const handleSubmit = useCallback(
    async (e: FormEvent) => {
      e.preventDefault();

      await RHFSubmit(async () => {
        try {
          await onSubmit(values as TFormValues);
          // eslint-disable-next-line @typescript-eslint/no-shadow
        } catch (e) {
          if (showSubmitError && isErrorLike(e) && isMounted.current) {
            setSubmitError({
              id: submitId,
              message: e.message ?? e,
            });
          }

          // eslint-disable-next-line no-console
          console.error(e);
        }
      })(e);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [RHFSubmit, onSubmit, values, isMounted, toggleSummaryFocus],
  );

  return (
    <FormIdContextProvider id={id}>
      <VStack
        width={width}
        id={id}
        onSubmit={handleSubmit}
        as="form"
        // align="start"
        spacing={5}
      >
        {showErrorSummary && (
          <ErrorSummary
            errors={allErrors}
            onFocus={toggleSummaryFocus.off}
            focusOnError={hasSummaryFocus}
          />
        )}

        {children}
      </VStack>
    </FormIdContextProvider>
  );
}
