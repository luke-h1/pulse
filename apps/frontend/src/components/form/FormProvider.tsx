import {
  FieldValues,
  FormProvider as RHFormProvider,
  useForm,
  UseFormProps,
  UseFormReturn,
} from 'react-hook-form';
import React, { ReactNode, useEffect, useRef } from 'react';
import { zodResolver } from '@hookform/resolvers/zod';
import z from 'zod';
import isEqual from 'lodash/isEqual';

interface FormProviderProps<TFormValues extends FieldValues> {
  children: ReactNode | ((form: UseFormReturn<TFormValues>) => ReactNode);
  enableReinitialize?: boolean;
  initialValues?: UseFormProps<TFormValues>['defaultValues'];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  validationSchema?: z.Schema<any, any>;
}

export default function FormProvider<TFormValues extends FieldValues>({
  children,
  enableReinitialize,
  initialValues,
  validationSchema,
}: FormProviderProps<TFormValues>) {
  const methods = useForm<TFormValues>({
    defaultValues: initialValues,
    mode: 'onBlur',
    resolver: validationSchema ? zodResolver(validationSchema) : undefined,
    shouldFocusError: false,
  });

  const previousInitialValues = useRef(initialValues);

  /**
   * RHF caches default values and doesn't have a built in option
   * to reinitialise forms.
   * This use effect provides the equivalent functionality as
   * Formik's `enableReinitialize` by checking for changes to the
   * initialValues and reseting the form with the new values if they've changed.
   */

  useEffect(() => {
    if (
      enableReinitialize &&
      !isEqual(previousInitialValues.current, initialValues)
    ) {
      methods.reset(initialValues as TFormValues);
      previousInitialValues.current = initialValues;
    }
  }, [enableReinitialize, methods, initialValues]);

  return (
    <RHFormProvider {...methods}>
      {typeof children === 'function' ? children(methods) : children}
    </RHFormProvider>
  );
}
