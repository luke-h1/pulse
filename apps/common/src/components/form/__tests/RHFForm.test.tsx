import z from 'zod';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FormProvider from '../FormProvider';
import RHFForm from '../RHFForm';

describe('Form', () => {
  test('Does not render errors for fields that do not have errors', async () => {
    const { container } = render(
      <FormProvider
        initialValues={{
          firstName: 'FirstName',
          lastName: '',
        }}
        validationSchema={z.object({
          firstName: z.string({
            required_error: 'First name is required',
          }),
          lastName: z.string({
            required_error: 'Last name is required',
          }),
        })}
      >
        <RHFForm id="test-form" onSubmit={jest.fn()}>
          Form
          <button type="submit">submit</button>
        </RHFForm>
      </FormProvider>,
    );

    userEvent.click(screen.getByRole('button', { name: 'submit' }));

    await waitFor(() => {
      expect(
        screen.queryByText('First name is required'),
      ).not.toBeInTheDocument();
    });

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('calls `onSubmit` handler when form is submitted successfully', async () => {
    const handleSubmit = jest.fn();

    render(
      <FormProvider
        initialValues={{
          firstName: 'firstName',
        }}
        validationSchema={z.object({
          firstName: z.string(),
        })}
      >
        <RHFForm id="test-form" onSubmit={handleSubmit}>
          Form
          <button type="submit">submit</button>
        </RHFForm>
      </FormProvider>,
    );

    userEvent.click(screen.getByRole('button', { name: 'submit' }));

    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
