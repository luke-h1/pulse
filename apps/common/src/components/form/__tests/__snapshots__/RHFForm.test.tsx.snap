// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Does not render errors for fields that do not have errors 1`] = `
<form class="chakra-stack css-22bnw2"
      id="test-form"
>
  <div class="chakra-stack css-1pyx7hi">
  </div>
  <button type="submit">
    submit
  </button>
</form>
`;
